Modernize Frame Component, propagate iframe globals (window, document) as context values

darthtrevino opened the pull request:
This PR is a refresh to update the component to modern JavaScript idioms. The components are expressed in JSX, and are using latest ES features and stage-2 features. Transpiled versions are available in the `lib/` folder.

Also added is a `DocumentContext` component that injects the iframe `document` and `window` into child components as a React context variable. (This was needed to get interop with `react-dnd` working (see https://github.com/gaearon/react-dnd/issues/241 and attached PRs).

The build system has been updated a bit. Gulp has been removed in favor of a handful of NPM scripts that execute linting, transpiling, and testing.

ryanseddon commented:
nice this has been on my todo list to make these changes, i'll give it a good review when I get a chance in a few days.

ryanseddon commented on .eslintrc:
```diff
@@ -0,0 +1,19 @@
+{
+  "extends": "airbnb",
```
not sure if this has leaked through from this preset but I'm not a fan of requiring a dangling comma.

darthtrevino replied:
Yeah, that one throws some people. I'll switch that rule around to enforce no dangling commas tomorrow morning.

ryanseddon commented on src/DocumentContext.jsx:
```diff
@@ -0,0 +1,27 @@
+import React, { Children, PropTypes } from 'react';
+
+export default class DocumentContext extends React.Component {
```
Destructure `Component` here too

ryanseddon replied:
tests for this component too?

darthtrevino commented on src/Frame.jsx:
```diff
@@ -0,0 +1,142 @@
+    contentDidUpdate: () => {},
+  };
+
+  static displayname = 'Frame';
```
I wasn't aware of that. There might be a babel transform that I can include. I'll research this in the morning.

jdreesen commented on src/Frame.jsx:
```diff
@@ -0,0 +1,142 @@
+  static displayName = 'Frame';
+
+  static defaultProps = {
+    initialContent: '<!DOCTYPE html><html><head></head><body><div class="frame-root"></div></body></html>',
```
I think this should be merged with the `defaultProps` above.

darthtrevino replied:
You'r right, the second defaultProps map is overriding the first.

darthtrevino commented:
@ryanseddon I think I've addressed your concerns

I briefly tried out enzyme to clean up the tests, but it didn't support iframes very well. They appear to be an edge case with most tooling ¯\_(ツ)_/¯

ryanseddon commented:
Thanks a lot for the incredible PR, I've added you as a collaborator 🙌
